File: lib/hologram/skuld/runner.ex
defmodule Hologram.Skuld.Runner do
  @moduledoc """
  Hologram-specific runner that wraps the ForeignResolver.Runner result
  in a `%Task{}` struct so it can be used with `Task.await/1` in actions.

      result =
        comp do
          h <- FiberPool.fiber(JS.call(:fetch, ["/api"]))
          FiberPool.await!(h)
        end
        |> JS.with_handler()
        |> FiberPool.with_handler()
        |> Hologram.Skuld.Runner.run()
        |> Task.await()

      put_state(component, :result, result)

  On the Hologram JS runtime, `run/1` calls `ForeignResolver.Runner.run/1`
  and wraps the resulting Promise in a `%Task{}` via `PromiseRegistry`.
  On the server, this raises.
  """

  alias Skuld.Comp.Types

  @doc """
  Run a computation through the ForeignResolver and wrap the result
  as a Task for use with `Task.await/1`.
  """
  @spec run(Types.computation()) :: Task.t()
  def run(comp) do
    _ensure_call_graph = &Skuld.ForeignResolver.Runner.run/1
    Hologram.Skuld.Runner.__wrap__(comp)
  end

  @doc false
  @spec __wrap__(Types.computation()) :: term()
  def __wrap__(_comp) do
    raise "Hologram.Skuld.Runner.__wrap__/1 is only available in the Hologram browser runtime"
  end
end

File: assets/js/elixir/hologram/skuld/runner.mjs
"use strict";

import ERTS from "../../../erts.mjs";
import Type from "../../../type.mjs";

const Elixir_Hologram_Skuld_Runner = {
  "__wrap__/1": (comp) => {
    const runner = globalThis.Elixir_Skuld_ForeignResolver_Runner;
    if (!runner) {
      throw new Error("Skuld.ForeignResolver.Runner module not loaded");
    }

    const result = runner["run/1"](comp);

    if (result instanceof Promise) {
      return ERTS.registerPromise(result);
    }

    return ERTS.registerPromise(Promise.resolve(result));
  },
};

export default Elixir_Hologram_Skuld_Runner;
